### PR TITLE
Feat/use sc addr in replacements

### PR DESCRIPTION
**Pull request by kilativ-dotcom**

### CODE REVIEW

File: problem-solver/cxx/inferenceModule/generator/SolutionTreeGenerator.hpp [hunk 1]
```diff
@@ -11,6 +11,8 @@
 
 #include <sc-memory/kpm/sc_agent.hpp>
 
+#include "utils/ReplacementsUtils.hpp"
```
Comment (MksmOrlov) [inline problem-solver/cxx/inferenceModule/generator/SolutionTreeGenerator.hpp:14]: You don't need it here

File: problem-solver/cxx/inferenceModule/searcher/templateSearcher/TemplateSearcherAbstract.cpp [hunk 2]
```diff
@@ -28,51 +28,40 @@ ScAddrVector TemplateSearcherAbstract::getInputStructures() const
...
-    element = formulaVariablesIterator->Get(2);
-    // TODO(MksmOrlov): replace with ScType::Var after new memory realisation
-    if (context->GetElementType(element) == ScType::NodeVar || context->GetElementType(element) == ScType::LinkVar)
-    {
-      variableSystemIdtf = context->HelperGetSystemIdtf(element);
-      if (!variableSystemIdtf.empty())
-        varNames.insert(variableSystemIdtf);
-    }
-  }
+    variables.insert(formulaVariablesIterator->Get(2));
```
Comment (MksmOrlov) [inline problem-solver/cxx/inferenceModule/searcher/templateSearcher/TemplateSearcherAbstract.cpp:64]: Why don't you add variables only with identifiers? Do you need unnamed variables?

File: problem-solver/cxx/inferenceModule/searcher/solutionTreeSearcher/SolutionTreeSearcher.cpp [hunk 3]
```diff
@@ -34,14 +34,13 @@ bool SolutionTreeSearcher::checkIfSolutionNodeExists(
...
+  for (ScAddr const & variable : variables)
   {
     ScAddr replacement;
-    templateParams.Get(varName, replacement);
+    templateParams.Get(variable, replacement);
     if (replacement.IsValid())
     {
-      ScAddr const & varNode = context->HelperFindBySystemIdtf(varName);
-      std::string pair_alias = replacement_for_alias + varName;
+      std::string pair_alias = replacement_for_alias + to_string(variable.Hash());
```
Comment (MksmOrlov) [inline problem-solver/cxx/inferenceModule/searcher/solutionTreeSearcher/SolutionTreeSearcher.cpp:43]: use camelCase

File: problem-solver/cxx/inferenceModule/searcher/templateSearcher/TemplateSearcherAbstract.cpp [hunk 4]
```diff
@@ -28,51 +28,40 @@ ScAddrVector TemplateSearcherAbstract::getInputStructures() const
...
 }
 
-void TemplateSearcherAbstract::getVarNames(ScAddr const & formula, std::set<std::string> & varNames)
+void TemplateSearcherAbstract::getVariables(ScAddr const & formula, ScAddrHashSet & variables)
 {
   ScIterator3Ptr const & formulaVariablesIterator =
-      context->Iterator3(formula, ScType::EdgeAccessConstPosPerm, ScType::Unknown);
-  ScAddr element;
-  std::string variableSystemIdtf;
+      context->Iterator3(formula, ScType::EdgeAccessConstPosPerm, ScType::Var);
```
Comment (MksmOrlov) [inline problem-solver/cxx/inferenceModule/searcher/templateSearcher/TemplateSearcherAbstract.cpp:62]: you can use iterator utils get all with type to get the same result

File: problem-solver/cxx/inferenceModule/manager/templateManager/TemplateManager.cpp [hunk 5]
```diff
@@ -21,15 +21,14 @@ TemplateManager::TemplateManager(ScMemoryContext * ms_context)
...
+  std::map<ScAddr, std::set<ScAddr, ScAddrLessFunc>, ScAddrLessFunc> replacementsMultimap;
   std::vector<ScTemplateParams> templateParamsVector;
 
   ScIterator3Ptr varIterator = context->Iterator3(scTemplate, ScType::EdgeAccessConstPosPerm, ScType::NodeVar);
   while (varIterator->Next())
   {
-    ScAddr var = varIterator->Get(2);
-    std::string varName = context->HelperGetSystemIdtf(var);
-    if (!replacementsMultimap[varName].empty())
+    ScAddr const & var = varIterator->Get(2);
```
Comment (MksmOrlov) [inline problem-solver/cxx/inferenceModule/manager/templateManager/TemplateManager.cpp:30]: Declare variable outside the cycle and name it as `variable`

**Review by MksmOrlov — CHANGES_REQUESTED** [review]
Write about replacements in scn-docunentation

File: problem-solver/cxx/inferenceModule/searcher/templateSearcher/TemplateSearcherAbstract.cpp (hunk 2 above)
Comment (kilativ-dotcom) [reply problem-solver/cxx/inferenceModule/searcher/templateSearcher/TemplateSearcherAbstract.cpp:64]: there may be formulas that have same unnamed variable in different atomic logical formulas

File: problem-solver/cxx/inferenceModule/searcher/templateSearcher/TemplateSearcherAbstract.cpp (hunk 4 above)
Comment (kilativ-dotcom) [reply problem-solver/cxx/inferenceModule/searcher/templateSearcher/TemplateSearcherAbstract.cpp:62]: I can do that, but getAllWithType creates temporary vector and insertions in this vector take time+memory

File: problem-solver/cxx/inferenceModule/manager/templateManager/TemplateManager.cpp (hunk 5 above)
Comment (kilativ-dotcom) [reply problem-solver/cxx/inferenceModule/manager/templateManager/TemplateManager.cpp:30]: I think it's better to declare this variable as const inside the loop. there may be some compiler optimizations for const variables

File: docs/scl-machine.tex [hunk 6]
```diff
@@ -186,6 +186,12 @@
     Если не нужно, то в процессе логического вывода используется \scnkeyword{пустой менеджер дерева решений}.}
 \end{scnindent}
 
+\scnheader{подстановки}
+\scnidtf{Replacements}
+\scntext{примечание}{Является однозначным соответствием между множеством sc-переменных и множеством кортежей sc-констант. Областью отправления соответствия является множество всех sc-переменных атомарной логической формулы, а соответствующим определённой sc-переменной элементом из области прибытия соответствия является кортеж, элементы которого при подстановке их в атомарную логическую формулу на место этой sc-переменнной делают формулу истинной. Корректно составленные подстановки имеют равномощные множества в области прибытия. Таким образом, элементы области прибытия могут образовывать прямоугольную матрицу. В этой матрице по колонкам для каждой переменной хранятся подстановки, при замене каждой sc-переменной на которую атомарная логическая формула становится истинной.}
```
Comment (MksmOrlov) [inline docs/scl-machine.tex:191]:
```suggestion
\scntext{примечание}{Является взаимно однозначным соответствием между множеством sc-переменных и множеством кортежей sc-констант. Областью отправления соответствия является множество всех sc-переменных атомарной логической формулы, а соответствующим определённой sc-переменной элементом из области прибытия соответствия является кортеж, элементы которого при подстановке их в атомарную логическую формулу на место этой sc-переменнной делают формулу истинной. Корректно составленные подстановки имеют равномощные множества в области прибытия. Таким образом, элементы области прибытия могут образовывать прямоугольную матрицу. В этой матрице по колонкам для каждой переменной хранятся подстановки, при замене каждой sc-переменной на которую атомарная логическая формула становится истинной.}
```

**Review by MksmOrlov — APPROVED** [review]
Check unresolved comment

**Pull request merged**